kale/pipeline/mida_trainer.py

```diff
@@ -0,0 +1,201 @@
+            x_transformed = self.transformer.fit_transform(x)
+            self._fit_mida(x_transformed, y, groups, transformer_params)
+
+        self.best_transformer = self.transformer.set_params(**self.best_transformer_params)
+        self.best_mida = MIDA(**self.best_mida_params)
+        self.best_estimator = self.estimator.set_params(**self.best_estimator_params)
+
+        x_transformed = self.best_transformer.transform(x)
+        x_transformed = self.best_mida.fit_transform(x_transformed, y=y, groups=groups)
+        self.best_estimator.fit(x_transformed[y.shape[0] :,], y)
+
+        return self
+
+    def _fit_mida(self, x_transformed, y, groups, transformer_params):
+        """Fit MIDA and estimator with the given data x_transformed and labels
+
+        Args:
+            x_transformed (array-like): Transformed data, shape (n_samples, n_features)
+            y (array-like): Data labels, shape (n_labeled_samples,)
+            groups (array-like): Group labels for the samples, shape (n_samples,)
+            transformer_params (dict): Parameters of the transformer
+
+        Returns:
+            self
+        """
+        self.best_mida = MIDA()
+        if self.mida_param_grid is None:
+            self.mida_param_grid = {key: [val] for key, val in self.best_mida.get_params().items()}
+        mida_grid = ParameterGrid(self.mida_param_grid)
+        for mida_params in mida_grid:
+            mida = MIDA(**mida_params)
+            x_transformed = mida.fit_transform(x_transformed, y=y, groups=groups)
+            cv = 5
+            if np.unique(groups).shape[0] > 2:
+                cv = LeaveOneGroupOut().split(x_transformed[y.shape[0] :], y, groups[y.shape[0] :])
+            self._fit_estimator(x_transformed, y, mida_params, transformer_params, cv=cv)
+
+        return self
+
+    def _fit_estimator(self, x_transformed, y, mida_params, transformer_params, cv=5):
+        """Fit the estimator with the given data x_transformed and labels y
+
+        Args:
+            x_transformed (array-like): Transformed data, shape (n_samples, n_features)
+            y (array-like): Data labels, shape (n_labeled_samples,)
+            mida_params (dict): Hyperparameters of the MIDA
+            transformer_params (dict): Hyperparameters of the transformer
+
+        Returns:
+            self
+        """
+        if self.estimator_param_grid is None:
+            self.estimator_param_grid = {key: [val] for key, val in self.estimator.get_params().items()}
+
+        clf = GridSearchCV(self.estimator, self.estimator_param_grid, cv=cv)
+        clf.fit(x_transformed[y.shape[0] :,], y)
+        if self.best_score is None or clf.best_score_ > self.best_score:
+            self.best_score = clf.best_score_
+            self.best_estimator_params = clf.best_params_
+            self.best_mida_params = mida_params
+            self.best_transformer_params = transformer_params
+
+        return self
+
+    def predict(self, x, groups=None):
```

Comment:
This could be useful to add to handle `predict`, `transform`, `predict_proba`, etc. 

`available_if`: https://scikit-learn.org/stable/modules/generated/sklearn.utils.metaestimators.available_if.html